### PR TITLE
[Merged by Bors] - feat: MonoidalCategory (QuadraticModuleCat R)

### PR DESCRIPTION
**Pull request by eric-wieser**
Together with #7130, this now results in a `CommMonoid (Skeleton (QuadraticModuleCat R))`, which is similar to the multiplicative part of the [Grothendieck-Witt ring](https://en.wikipedia.org/wiki/Witt_group#Grothendieck-Witt_ring) of quadratic forms.

(Strictly the Grothendieck-Witt ring would introduce a subcategory of QuadraticModuleCat for the non-singular forms)

This is largely a copy-and-paste from #7598 (though it was written first)

---

- [x] depends on: #7237
- [x] depends on: #7387

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Lean/Meta/DiscrTree.lean [hunk 1]
```diff
@@ -53,19 +53,25 @@ partial def getSubexpressionMatches (d : DiscrTree α s) (e : Expr) : MetaM (Arr
...
+  | empty : Ctxt α β γ δ
+  | ctxt : α → β → γ → δ → Ctxt α β γ δ → Ctxt α β γ δ
+
+/-- Apply a function to the array of values at each node in a `DiscrTree`. -/
+partial def Trie.mapArrays (t : Trie α s) (f : Array α → Array β) : Trie β s :=
+  let .node vs0 cs0 := t
+  go Ctxt.empty #[] (f vs0) cs0.toList
+where
+  /- This implementation as a single tail-recursive function is chosen to not blow the
+     interpreter stack when the `Trie` is very deep -/
```
Comment (nomeata) [inline Mathlib/Lean/Meta/DiscrTree.lean:68]:
```suggestion
  /-- This implementation as a single tail-recursive function is chosen to not blow the
      interpreter stack when the `Trie` is very deep -/
```

**Comment by eric-wieser** [comment]
@joelriou, I requested a review since this is almost identical to #7598. I believe I already applied your suggestions there, but it's possible I missed something.

File: Mathlib/LinearAlgebra/QuadraticForm/QuadraticModuleCat/Monoidal.lean [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        erw [Category.id_comp, Category.comp_id, MonoidalCategory.tensor_id, Category.comp_id]
+        rfl
+      leftUnitor := fun X => ofIso (tensorLId X.form)
+      rightUnitor := fun X => ofIso (tensorRId X.form)
+      -- uncomment these lines to avoid running slow proofs while editing the proof above
+      -- whiskerLeft_eq := sorry
+      -- whiskerRight_eq := sorry
+      -- tensorHom_eq := sorry
+      -- leftUnitor_eq := sorry
+      -- rightUnitor_eq := sorry
```
Comment (joelriou) [inline Mathlib/LinearAlgebra/QuadraticForm/QuadraticModuleCat/Monoidal.lean:98]: This should be removed before merging.

**Comment by joelriou** [comment]
Could you clarify the PR toplevel comment? (It seems there is a numbering issue with PR 7215.)

**Comment by eric-wieser** [comment]
Whoops, good catch. Fixed.

**Comment by joelriou** [comment]
bors merge

**Pull request closed without merging**